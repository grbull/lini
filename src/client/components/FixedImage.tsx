import cn from 'classnames';
import React, { ReactElement, useEffect, useState } from 'react';

import { Skeleton } from './Skeleton';

interface Props {
  alt: string;
  className?: string;
  fullHeight: number;
  fullWidth: number;
  skeletonClassName?: string;
  src: string;
}

export function FixedImage({
  alt,
  className,
  fullHeight,
  fullWidth,
  skeletonClassName,
  src,
}: Props): ReactElement {
  const [isLoading, setIsLoading] = useState(false);
  const [imageHeightPercentage, setImageHeightPercentage] = useState(
    (fullHeight / fullWidth) * 100
  );

  useEffect(() => {
    const imageElement = new Image();
    imageElement.onload = (): void => {
      if (!isLoading) {
        setIsLoading(false);
        setImageHeightPercentage(
          (imageElement.naturalHeight / imageElement.naturalWidth) * 100
        );
      }
    };
    imageElement.src = src;
  }, [src, isLoading]);

  return (
    <div
      className={cn('h-0', 'relative', className)}
      style={{ paddingTop: `${imageHeightPercentage}%` }}
    >
      <img
        alt={alt}
        className={cn('absolute', 'top-0', 'left-0', {
          hidden: isLoading,
        })}
        src={src}
      />
      <Skeleton
        className={cn('absolute', 'top-0', 'left-0', skeletonClassName, {
          hidden: !isLoading,
        })}
      />
    </div>
  );
}

FixedImage.defaultProps = {
  className: undefined,
  skeletonClassName: undefined,
};
