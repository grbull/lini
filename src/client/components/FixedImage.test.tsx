/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import React from 'react';

import { FixedImage } from './FixedImage';

describe('FixedImage Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <FixedImage
        alt="image"
        fullHeight={1080}
        fullWidth={1920}
        src="/image.png"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
