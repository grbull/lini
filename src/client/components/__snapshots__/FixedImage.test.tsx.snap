// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedImage Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-0 relative"
    style="padding-top: 56.25%;"
  >
    <div
      class="animate-pulse bg-gray-700 absolute top-0 left-0 w-full h-full"
    />
    <img
      alt="image"
      class="absolute top-0 left-0 w-full h-full hidden"
      src="/image.png"
    />
  </div>
</DocumentFragment>
`;
